examples/ode-integrator.dx

```diff
@@ -1,60 +1,179 @@
+def length (x: d=>Float) : Float = sqrt $ sum for i. sq x.i
+def (./) (x: d=>Float) (y: d=>Float) : d=>Float = for i. x.i / y.i
+def dot (_:VSpace v) ?=> (s:d=>Float) (vs:d=>v) : v = sum for j. s.j .* vs.j
+
+def fit_4th_order_polynomial (_:VSpace v) ?=>
+    (z0:v) (z1:v) (z_mid:v) (dz0:v) (dz1:v) (dt:Time) : (Fin 5)=>v =
+  -- dz0 and dz1 are gradient evaluations.
+  a = -2. * dt .* dz0 + 2. * dt .* dz1 -  8. .* z0 -  8. .* z1 + 16. .* z_mid
+  b =  5. * dt .* dz0 - 3. * dt .* dz1 + 18. .* z0 + 14. .* z1 - 32. .* z_mid
+  c = -4. * dt .* dz0 +      dt .* dz1 - 11. .* z0 -  5. .* z1 + 16. .* z_mid
+  d = dt .* dz0
+  e = z0
+  [a, b, c, d, e]  -- polynomial coefficients.
+
+dps_c_mid = [6025192743. /30085553152. /2., 0., 51252292925. /65400821598. /2.,
+            -2691868925. /45128329728. /2., 187940372067. /1594534317056. /2.,
+            -1776094331. /19743644256. /2., 11237099. /235043384. /2.]
+
+def interp_fit_dopri (_:VSpace v) ?=>
+    (z0:v) (z1:v) (k:(Fin 7)=>v) (dt:Time) : (Fin 5)=>v =
+  -- Fit a polynomial to the results of a Runge-Kutta step.
+  z_mid = z0 + dt .* (dot dps_c_mid k)
+  fit_4th_order_polynomial z0 z1 z_mid k.(0@_) k.(1@_) dt
+
+def initial_step_size (fun:d=>Float -> Time -> d=>Float) (t0:Time) (z0:d=>Float)
+                      (order:Int) (rtol:Float) (atol:Float) (f0:d=>Float) : Time =
+  -- Algorithm from: E. Hairer, S. P. Norsett G. Wanner,
+  -- Solving Ordinary Differential Equations I: Nonstiff Problems, Sec. II.4.
+  scale = for i. atol + ((abs z0.i) * rtol)
+  d0 = length (z0 ./ scale)
+  d1 = length (f0 ./ scale)
+  h0 = select ((d0 < 1.0e-5) || (d1 < 1.0e-5)) 1.0e-6 (0.01 * (d0 / d1))
+  z1 = z0 + h0 .* f0
+  f1 = fun z1 (t0 + h0)
+  d2 = (length ((f1 - f0) ./ scale)) / h0
+  left = max 1.0e-6 (h0 * 1.0e-3)
+  right = pow (0.01 / (d1 + d2)) (1. / ((IToF order) + 1.))
+  h1 = select ((d1 <= 1.0e-15) && (d2 <= 1.0e-15)) left right
+  min (100. * h0) h1
+
+-- Dopri5 Butcher tableaux
+alpha = [1. / 5., 3. / 10., 4. / 5., 8. / 9., 1., 1.]
+
+beta : (i:(Fin 6)=>(..i)=>Float) =  -- triangular array type.
+       [[1. / 5.],
+       [3. / 40., 9. / 40.],
+       [44. / 45., -56. / 15., 32.0 / 9.],
+       [19372. / 6561., -25360. / 2187., 64448. / 6561., -212. / 729.],
+       [9017./3168., -355./33., 46732./5247., 49./176., -5103./18656.],
+       [35. / 384., 0., 500. / 1113., 125./192., -2187./6784., 11./84.]]
+
+c_sol = [35. /384., 0., 500. /1113., 125. /192., -2187. /6784., 11. / 84., 0.]
+c_error = [35. / 384. - 1951. / 21600., 0., 500. / 1113. - 22642. / 50085.,
+           125. / 192. - 451. / 720., -2187. / 6784. + 12231. / 42400.,
+           11. / 84. - 649. / 6300., -1. / 60.]
+
+def runge_kutta_step (_:VSpace v) ?=> (func:v->Time->v)
+    (z0:v) (f0:v) (t0:Time) (dt:Time) : (v & v & v & (Fin 7)=>v) =
+
+  runge_kutta_stage = \i func_evals.
+    ti = t0 + dt .* alpha.i
+    zi = z0 + dt .* dot beta.i func_evals
+    func zi ti
+
+  evals_init = for i. select (i == (0@_)) f0 zero
```

Comment:
Wouldn't it be simpler to initialize the state with `zero`s and then set the first index to `f0`?

Reply:
Here's the simplest way I know how to write that:
```
evals_init = snd $ withState zero \r.
  r!(0@_) := f0
```
I'll put that version in for now, although it's not all that readable either imo.

Reply:
Oh, I meant fusing the assignment into the subsequent `withState` block. So instead of doing
```
evals_init = for i. select (i == (0@_)) f0 zero
withState evals_init \evals.
  ...
```
you would do
```
withState zero \evals.
  evals!(0@_) := f0
  ...
```
which I think is nicer?

Reply:
Oh, of course!  Yes, I'll fix that next time I touch that code.

It's funny, because I learned how to use ```withState``` just by looking at examples, it didn't occur to me until this conversation that you could decouple it from a ```for``` loop.  I was thinking of it as just something you add to a for loop to let it make mutating updates.

Reply:
Right, you can definitely use `withState` without loops, although it usually doesn't make sense, because if you have the code unrolled then you don't need the state effect! But it's good to be able to e.g. do this initialization outside of the loop.